### DISCUSSION TITLE
come controllare scadenza certificato tramite HA su docker

### DISCUSSION THREAD

**Discussion by poliant** [indomus/forum#1202] [GENERALE - Discussioni generiche]
ciao a tutti,
ho un raspi4 dove girano un po' di servizi per la mia domotica (home assistant, duckdns, caddy, etc.). Praticamente tutti i servizi li ho su container docker (un container per HA, uno per duckdns, uno per caddy come reverse proxy, etc.).

Ed ecco il problema: l'integrazione "certificate expiry" mi da che il certificato che ho installato (tramite caddy su docker) è non valido: è perché il certificato risiede su un container differente da quello di HA e quindi HA non ha accesso al certificato?

grazie dell'aiuto,
Antonio

PS: ovviamente il certificato è valido ed ho controllato la validità sia direttamente nel certificato tramite `ssl-cert-check` che tramite i vari siti che controllano la validità dei certificati ssl.

**Comment by snakuzzo** [comment]
Quell'integrazione non legge i certificati su disco (che tra l'altro sono nel container di caddy...e non di HA). Semplciemente verifica che "contattando la tua URL" venga restituito un certificato valido.

**Reply by poliant** [reply]
ciao Diego,
certamente HA non ha accesso al certificato che sta su caddy e proprio per questo motivo pensavo mi desse non valido o meglio "unknown".
Però stando a quanto dici, allora la cosa strana è che HA mi dice che il certificato è self-signed (sotto)
![image](https://user-images.githubusercontent.com/26714536/146535582-fb6fd141-9438-4630-a759-47737f9daba4.png)

Però il certificato io l'ho generato tramite caddy e controllado tramite vari siti di check (ssllabs.com, sslshopper.com) risulta valido - e questi siti dovrebbero fare stesso controllo dell'integrazione di HA.
da ssl labs:
![image](https://user-images.githubusercontent.com/26714536/146536764-e00ebdb4-9dad-4a4e-a032-48e218e62eaf.png)

da sslshopper
![image](https://user-images.githubusercontent.com/26714536/146537161-d5192153-f1ad-4309-b3d5-7ef569557e42.png)

Secondo te dove sta il problema?

grazie

**Reply by snakuzzo** [reply]
Guarda...è vero che hai mascherato il tuo FQDN, ma non l'hai fatto abbastanza bene. Per cui mi sono preso la libertà di fare una prova ed il tuo certificato risulta essere corretto. Non so su che porta hai esposto HA, ma puntando direttamente il tuo dominio sulla 443 il certificato risulta ok ed in scadenza il 15 marzo 2022.

Guarda...

![1](https://user-images.githubusercontent.com/7205371/146538735-991edb41-2906-4bbf-8888-c00df79e8e4c.png)

Quindi il certificato è ok e probabilmente hai indicato tu una URL sbagliata nell'integrazione che punta qualche altra cosa che ti espone un certificato self-signed. Più di questo non so dirti.
Se mi contatti in privato su telegram possiamo fare qualche prova insieme.

**Reply by poliant** [reply]
ciao Diego,
avevo mascherato (male....) perché ricevendo info strane da HA non volevo farlo vedere, ma va bene così che hai fatto il check :-)

HA è esposto su porta 9443 perché altrimenti avevo problemi con il termostato Netatmo e sulla 443 vi è il semplice messaggio di test.

Cmq il problema mi si presenta da quando ho cambiato provider della fibra: ora ho vodafone FTTH ed uso la vodafone station wifi 6 (prima avevo TIM FTTC con fritzbox 7590 - ora lo uso come AP WiFi visto che il WiFi della vodafone station è pessimo!!!).
Ovviamente sulla vodafone station ho aperto le porte che mi servono (e.g. 80, 443 e 9443) ed infatti provando dal cellulare collegato alla rete mobile riesco ad accedere sia ad HA che al messaggio su porta 443 e vengo reinstradato correttamente da http ad https e il certificato risulta essere valido.

Mi è venuto in mente di fare la prova di collegarmi al mio FQDN dalla lan e la cosa stranissima è che vengo reindirizzato al router... mentre mi aspettavo di essere reindirizzato al raspberry.

boh!?!?!?!?

**Reply by snakuzzo** [reply]
A me dà ovviamente ok anche puntando la porta 9443.
 
![1](https://user-images.githubusercontent.com/7205371/146546087-648ce5e6-44e0-4263-9b41-bce5014fa99f.png)

**Reply by poliant** [reply]
controllando sui fari forum relativi a vodafone ftth sembra che il problema sia la vodafone station wifi 6: ha molti bug e uno di questi è prioprio il mio problema: non suporta la NAT Reflection o NAT loopback!
Questa limitazione comporta che quando il rpi fa una richiesta al mio FQDN, la vodafone station risponde con una richiesta alla sua porta sulla LAN e quindi il check fallisce! Come anche fallisce il routing verso il rpi

**Reply by snakuzzo** [reply]
ok

**Reply by snakuzzo** [reply]
Quindi tu il problema di fatto ce l'hai solo quando sei nella tua LAN. Dall'esterno invece tutto ok (infatti io raggiungo correttamente sia la tua istanza di HA che di Caddy)

**Reply by poliant** [reply]
si!
ma sono impazzito per capire dove sta il problema.... incredibile che vodafone da un modem/router così scarso....

**Reply by manny199** [reply]
Anche io ho HA su docker con caddy ed il mio certiicato ha attualmente validità 3 mesi. Non mi è chiaro come verificare qualora per qualche strana ragione non si dovesse rinnovare automaticamente alla scadenza

**Reply by snakuzzo** [reply]
Ti basta abilitare l'integrazione "Cert Expiry" impostando ovviamente la URL con cui raggiungi da internet Home Assistant (es. `http://miodominio.duckdns.org:9000` che ti creerà un'entità di tipo sensor che ti restituirà un timestamp (es.  `sensor.cert_expiry_timestamp_miodominio_duckdns_org_9000`)

A questo punto puoi crearti semplicemente un template sensor così...

```yaml
sensor:

- platform: template
  sensors:
    ssl_days:
      friendly_name: "SSL"
      value_template: "{{ (( as_timestamp(states('sensor.cert_expiry_timestamp_miodominio_duckdns_org_9000'), '0') | int - as_timestamp(now()) | int ) /60/60/24 ) | int }}"
```

Al riavvio di HA ti ritroverai con un'entità di tipo sensor come questa che ti restituisce un numerico che corrisponde ai giorni che mancano alla scadenza (nel mio caso 34)...

![1](https://user-images.githubusercontent.com/7205371/147700972-0c6b0c3c-9779-4bcd-a1bb-5cc32997e1a5.png)

**Reply by manny199** [reply]
Ma i certificati SSL con caddy si aggiornano sempre automaticamente?

**Reply by snakuzzo** [reply]
se hai configurato tutto correttamente...sì
